pam.c

```diff
@@ -35,17 +33,41 @@ static int handle_conversation(int num_msg, const struct pam_message **msg,
 	}
 	*resp = pam_reply;
 	for (int i = 0; i < num_msg; ++i) {
+		swaylock_log(LOG_DEBUG, "PAM message %d: %s", i, msg[i]->msg);
 		switch (msg[i]->msg_style) {
-		case PAM_PROMPT_ECHO_OFF:
 		case PAM_PROMPT_ECHO_ON:
+			if(write_comm_text_message_from_backend(&msg[i]->msg) < 0) {
+				swaylock_log(LOG_ERROR, "Failed to write message from backend");
+				return PAM_ABORT;
+			}
+			#ifdef __GNUC__
+			// both PAM_PROMPT_ECHO_ON and PAM_PROMPT_ECHO_OFF expect a string response
+			__attribute__((fallthrough));
+			#endif
+
+		case PAM_PROMPT_ECHO_OFF:
+			char *pw_buf = NULL;
+			ssize_t size = read_comm_prompt_response(&pw_buf);
+			if (size < 0) {
+				swaylock_log(LOG_ERROR, "Failed to read prompt response");
+				return PAM_ABORT;
+			}
+			//TODO: what to do if size == 0? in fbc5a8136187, if read_comm_prompt_response (there called read_comm_request) returned 0, it would break the while loop (i.e. auth successfully); we can't do that here
```

Comment:
If `size==0`, then the pipe connecting to the main `swaylock` process has disconnected because e.g. swaylock has exited (e.g. unlocked by signal, or crashed). I think the correct thing to do here would be to return `PAM_ABORT`, so that `pam_authenticate` returns `PAM_ABORT`; the current authentication attempt cannot complete.

If you furthermore want `run_pw_backend_child()` to not warn about this specific exit path, you could provide it with additional information by passing a pointer to a variable declared in `run_pw_backend_child` as  `.appdata_ptr` in `struct pam_conv conv`, and having `handle_conversation` modify the variable depending on the reason for exit.

Reply:
Oh, you are right. I guess my thinking was that "`char *` is a string, and a string should be passed by reference instead of value". But of course it's already a reference because it's a pointer. I fixed this now in bc18cc2 (I can clean up the commit history at the end.)

I also see that `string[offs]` is wrong, but how come it works? ... Oh is it because `offs` always happens to be zero?

Reply:
Yeah, it makes sense to return `PAM_ABORT`. I think I was confused because in the [current version](https://github.com/swaywm/swaylock/blob/fbc5a8136187491e5ee3071e8a60e6593ca9b90b/pam.c#L95-L102), `size == 0` breaks the while loop and exits with `EXIT_SUCCESS`, so I thought it indicates success, but indeed the subprocess exit status doesn't necessarily say anything about success on the main process side.